### PR TITLE
python312Packages.skein: unbreak by patching distutils,

### PR DESCRIPTION
**Pull request by teutat3s**
## Description of changes
remove vendorized versioneer

ZHF: #309482

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by illustris** [comment]
```
[nix-shell:~/.cache/nixpkgs-review/pr-310218]$ nix-shell -I nixpkgs=$PWD/nixpkgs -p 'pkgs.python3.withPackages (p: with p; [ skein ])'
this derivation will be built:
  /nix/store/xav2y68ays9p0hqlsx1xximdl0hgiqqg-python3-3.11.9-env.drv
building '/nix/store/xav2y68ays9p0hqlsx1xximdl0hgiqqg-python3-3.11.9-env.drv'...
created 258 symlinks in user environment

[nix-shell:~/.cache/nixpkgs-review/pr-310218]$ python
Python 3.11.9 (main, Apr  2 2024, 08:25:04) [GCC 13.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import skein
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nix/store/4ix14p6233al1d94gs55bvc1c48d0cqi-python3.12-skein-0.8.2/lib/python3.12/site-packages/skein/__init__.py", line 1, in <module>
    from . import kv
  File "/nix/store/4ix14p6233al1d94gs55bvc1c48d0cqi-python3.12-skein-0.8.2/lib/python3.12/site-packages/skein/kv.py", line 12, in <module>
    import grpc as _grpc
  File "/nix/store/2qkc40wsir4qrynfwfmqlbssbcvaincq-python3.12-grpcio-1.62.2/lib/python3.12/site-packages/grpc/__init__.py", line 22, in <module>
    from grpc import _compression
  File "/nix/store/2qkc40wsir4qrynfwfmqlbssbcvaincq-python3.12-grpcio-1.62.2/lib/python3.12/site-packages/grpc/_compression.py", line 20, in <module>
    from grpc._cython import cygrpc
ImportError: cannot import name 'cygrpc' from 'grpc._cython' (/nix/store/2qkc40wsir4qrynfwfmqlbssbcvaincq-python3.12-grpcio-1.62.2/lib/python3.12/site-packages/grpc/_cython/__init__.py)
>>>

[nix-shell:~/.cache/nixpkgs-review/pr-310218]$
exit

[nix-shell:~/.cache/nixpkgs-review/pr-310218]$ nix-shell -I nixpkgs=$PWD/nixpkgs -p 'pkgs.python312.withPackages (p: with p; [ skein ])'

[nix-shell:~/.cache/nixpkgs-review/pr-310218]$ python
Python 3.12.3 (main, Apr  9 2024, 08:09:14) [GCC 13.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import skein
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nix/store/4ix14p6233al1d94gs55bvc1c48d0cqi-python3.12-skein-0.8.2/lib/python3.12/site-packages/skein/__init__.py", line 1, in <module>
    from . import kv
  File "/nix/store/4ix14p6233al1d94gs55bvc1c48d0cqi-python3.12-skein-0.8.2/lib/python3.12/site-packages/skein/kv.py", line 17, in <module>
    from .model import (
  File "/nix/store/4ix14p6233al1d94gs55bvc1c48d0cqi-python3.12-skein-0.8.2/lib/python3.12/site-packages/skein/model.py", line 13, in <module>
    from .objects import Enum, ProtobufMessage, Specification, required
  File "/nix/store/4ix14p6233al1d94gs55bvc1c48d0cqi-python3.12-skein-0.8.2/lib/python3.12/site-packages/skein/objects.py", line 7, in <module>
    from .utils import format_list, ensure_unicode, datetime_to_millis
  File "/nix/store/4ix14p6233al1d94gs55bvc1c48d0cqi-python3.12-skein-0.8.2/lib/python3.12/site-packages/skein/utils.py", line 9, in <module>
    from setuptools._distutils.version import LooseVersion
ModuleNotFoundError: No module named 'setuptools'
```

Build succeeds, but fails to import. I'm not sure how it's passing the `pythonImportsCheck = [ "skein" ];`. Looking into it.

File: pkgs/development/python-modules/skein/default.nix [hunk 1]
```diff
@@ -10,6 +11,7 @@
 , hadoop
 , pytestCheckHook
 , python
+, versioneer
```
Comment (illustris) [inline pkgs/development/python-modules/skein/default.nix:14]:
```suggestion
, setuptools
, versioneer
```

Comment (illustris) [reply pkgs/development/python-modules/skein/default.nix:14]: `setuptools` will need to be added to `propagatedBuildInputs` for python 3.12
```nix
propagatedBuildInputs = [ cryptography grpcio pyyaml ]
  ++ lib.optionals (!pythonOlder "3.12") [ setuptools ];
```

**Review by illustris — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/skein/default.nix (hunk 1 above)
Comment (teutat3s) [reply pkgs/development/python-modules/skein/default.nix:15]: Thanks for the review. Done.

**Pull request merged**